Sources/ContentScopeScripts/dist/contentScope.js

```diff
@@ -3864,78 +3864,230 @@
+
+      // apply styles to hide element
       element.style.setProperty('display', 'none', 'important');
+      element.style.setProperty('min-height', '0px', 'important');
+      element.style.setProperty('height', '0px', 'important');
       element.hidden = true;
   }
 
+  /**
+   * Show previously hidden DOM element
+   * @param {HTMLElement} element
+   */
+  function unhideNode (element) {
+      const cachedDisplayProperties = hiddenElements.get(element);
+      if (!cachedDisplayProperties) {
+          return
+      }
+
+      for (const prop in cachedDisplayProperties) {
+          element.style.setProperty(prop, cachedDisplayProperties[prop]);
+      }
+      hiddenElements.delete(element);
+      element.hidden = false;
+  }
+
+  /**
+   * Check if DOM element contains visible content
+   * @param {HTMLElement} node
+   */
   function isDomNodeEmpty (node) {
-      const visibleText = node.innerText.trim().toLocaleLowerCase();
-      const mediaContent = node.querySelector('video,canvas');
-      const frameElements = [...node.querySelectorAll('iframe')];
+      // no sense wasting cycles checking if the page's body element is empty
+      if (node.tagName === 'BODY') {
+          return false
+      }
+      // use a DOMParser to remove all metadata elements before checking if
+      // the node is empty.
+      const parsedNode = parser.parseFromString(node.outerHTML, 'text/html').documentElement;
+      parsedNode.querySelectorAll('base,link,meta,script,style,template,title,desc').forEach((el) => {
+          el.remove();
+      });
+
+      const visibleText = parsedNode.innerText.trim().toLocaleLowerCase().replace(/:$/, '');
+      const mediaContent = parsedNode.querySelector('video,canvas,picture');
+      const frameElements = [...parsedNode.querySelectorAll('iframe')];
       // about:blank iframes don't count as content, return true if:
       // - node doesn't contain any iframes
       // - node contains iframes, all of which are hidden or have src='about:blank'
       const noFramesWithContent = frameElements.every((frame) => {
           return (frame.hidden || frame.src === 'about:blank')
       });
+
       if ((visibleText === '' || adLabelStrings.includes(visibleText)) &&
           noFramesWithContent && mediaContent === null) {
           return true
       }
       return false
   }
 
-  function hideMatchingDomNodes (rules) {
+  /**
+   * Apply relevant hiding rules to page at set intervals
+   * @param {Object[]} rules
+   * @param {string} rules[].selector
+   * @param {string} rules[].type
+   */
+  function applyRules (rules) {
+      // several passes are made to hide & unhide elements. this is necessary because we're not using
+      // a mutation observer but we want to hide/unhide elements as soon as possible, and ads
+      // frequently take from several hundred milliseconds to several seconds to load
+      // check at 0ms, 100ms, 200ms, 300ms, 400ms, 500ms, 1000ms, 1500ms, 2000ms, 2500ms, 3000ms
```

Comment:
This may be excessive, but there are a few variables (connection speed, website speed, device speed) that impact when ad containers are added to the page. On high end devices with a reasonable connection speed to an average website, most ad containers are on the page well before 500ms. But if you have a slow device, a bad connection, or are loading a slow website, this can take upwards of 1s--I've been testing this feature under each of these conditions and the slowest case I've found was on indiatimes.com, where when navigating from page to page using the top nav the leaderboard container can take ~2.5s to be added to the page.